source/extensions/filters/common/local_ratelimit/provider_singleton.h

```diff
@@ -0,0 +1,106 @@
+#pragma once
+
+#include <functional>
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "envoy/type/v3/token_bucket.pb.h"
+#include "envoy/type/v3/token_bucket.pb.validate.h"
+
+#include "source/common/config/subscription_base.h"
+#include "source/common/init/target_impl.h"
+#include "source/extensions/filters/common/local_ratelimit/local_ratelimit_impl.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Filters {
+namespace Common {
+namespace LocalRateLimit {
+
+class RateLimiterProviderSingleton;
+using RateLimiterProviderSingletonSharedPtr = std::shared_ptr<RateLimiterProviderSingleton>;
+class RateLimiterProviderSingleton : public Singleton::Instance {
+public:
+  class RateLimiterWrapper {
+  public:
+    RateLimiterWrapper(RateLimiterProviderSingletonSharedPtr provider,
+                       std::shared_ptr<LocalRateLimiterImpl> limiter)
+        : provider_(provider), limiter_(limiter) {}
+    std::shared_ptr<LocalRateLimiterImpl> getLimiter() const { return limiter_; }
+    void setLimiter(std::shared_ptr<LocalRateLimiterImpl> limiter) { limiter_ = limiter; }
+
+  private:
+    // The `provider_` holds the ownership of this singleton by shared
+    // pointer, as the rate limiter map singleton isn't pinned and is
+    // shared among all the access log rate limit filters.
+    RateLimiterProviderSingletonSharedPtr provider_;
```

Comment:
I may be missing something, but why is this needed?
IIUC all the filter needs is the LocalRateLimiterImpl object.